### PR TITLE
Support Let/Set in autoskip

### PR DESCRIPTION
**Pull request by idodod**
Address https://github.com/earthly/earthly/issues/3996

### CODE REVIEW

File: tests/wildcard-copy.earth [hunk 1]
```diff
@@ -45,9 +45,10 @@ wildcard-multi-with-flags:
     RUN test -f somedir1/somefile1
 
 wildcard-if-exists:
-    #COPY --if-exists ./wildcard/*+test/non-existing .
-    #DO +TEST --EXPECTED_COUNT=0
-    #RUN ! test -e non-existing
+    COPY --if-exists ./wildcard/*+test/non-existing .
```
Comment (idodod) [inline tests/wildcard-copy.earth:48]: forgot to uncomment this test in previous PR

**Review by mikejholly — APPROVED** [review]
Nice work!

File: inputgraph/loader.go [hunk 2]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
 }
 
+func (l *loader) handleLet(ctx context.Context, cmd spec.Command) error {
+	var opts commandflag.LetOpts
+	argsCpy := flagutil.GetArgsCopy(cmd)
+	args, err := flagutil.ParseArgsCleaned("LET", &opts, argsCpy)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse LET args")
+	}
+	if len(args) != 3 {
```
Comment (alexcb) [inline inputgraph/loader.go:468]: should we also check for a `=` somewhere? currently we'll accept any of `LET foo <- bar`, `LET foo gets bar`, or `LET foo = bar`
```suggestion
	if len(args) != 3 || args[1] != '=' {
```

File: inputgraph/loader.go [hunk 3]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+	if err != nil {
+		return wrapError(err, cmd.SourceLocation, "failed to expand LET value %q", baseVal)
+	}
+
+	l.hasher.HashString(fmt.Sprintf("LET %s=%s", key, val))
+
+	_, _, err = l.varCollection.DeclareVar(key, variables.WithValue(val))
+	if err != nil {
+
+		if err != nil {
```
Comment (alexcb) [inline inputgraph/loader.go:483]: you must really want to check err isn't null here :)

(there's a duplicate if above)

File: inputgraph/loader.go [hunk 4]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+			return wrapError(err, cmd.SourceLocation, "failed to declare variable")
+		}
+	}
+	return nil
+}
+
+func (l *loader) handleSet(ctx context.Context, cmd spec.Command) error {
+	var opts commandflag.SetOpts
+	argsCpy := flagutil.GetArgsCopy(cmd)
+	args, err := flagutil.ParseArgsCleaned("LET", &opts, argsCpy)
```
Comment (alexcb) [inline inputgraph/loader.go:493]:
```suggestion
	args, err := flagutil.ParseArgsCleaned("SET", &opts, argsCpy)
```

File: inputgraph/loader.go [hunk 5]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
 
+func (l *loader) handleLet(ctx context.Context, cmd spec.Command) error {
+	var opts commandflag.LetOpts
+	argsCpy := flagutil.GetArgsCopy(cmd)
+	args, err := flagutil.ParseArgsCleaned("LET", &opts, argsCpy)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse LET args")
+	}
+	if len(args) != 3 {
+		return errors.New("failed to parse LET args")
```
Comment (alexcb) [inline inputgraph/loader.go:469]: not sure if this is actually the syntax that we need here, but maybe there's a way to include the source location for this error too?
```suggestion
		return wrapError(errors.New("failed to parse LET args"), cmd.SourceLocation)
```

File: inputgraph/loader.go (hunk 2 above)
Comment (idodod) [reply inputgraph/loader.go:468]: The reason for only checking length is that I didn't want to potentially get an index out of bound error. 
I think the rest of the validation should still occur later on when LET/SET are handled by the interperter.
Thoughts?

File: inputgraph/loader.go (hunk 5 above)
Comment (alexcb) [reply inputgraph/loader.go:469]: actually.....

```
return newError(with.Command.SourceLocation,"failed to parse LET args")
```

Comment (alexcb) [reply inputgraph/loader.go:469]: there's a number of places in this PR where we could do this.

File: inputgraph/loader.go [hunk 6]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+	}
+	return nil
+}
+
+func (l *loader) handleSet(ctx context.Context, cmd spec.Command) error {
+	var opts commandflag.SetOpts
+	argsCpy := flagutil.GetArgsCopy(cmd)
+	args, err := flagutil.ParseArgsCleaned("LET", &opts, argsCpy)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse LET args")
```
Comment (alexcb) [inline inputgraph/loader.go:495]:
```suggestion
		return newError(with.Command.SourceLocation,"failed to parse SET args")
```

File: inputgraph/loader.go [hunk 7]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+	}
+	if len(args) != 3 {
+		return errors.New("failed to parse LET args")
```
Comment (alexcb) [inline inputgraph/loader.go:498]: let -> set

File: inputgraph/loader.go [hunk 8]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+	}
+
+	key := args[0]
+	baseVal := args[2]
+	val, err := l.expandArgs(ctx, baseVal)
+	if err != nil {
+		return wrapError(err, cmd.SourceLocation, "failed to expand LET value %q", baseVal)
```
Comment (alexcb) [inline inputgraph/loader.go:505]:
```suggestion
		return wrapError(err, cmd.SourceLocation, "failed to expand SET value %q", baseVal)
```

**Review by alexcb — COMMENTED** [review]
looking good; just a number of small suggestions.

File: inputgraph/loader.go (hunk 2 above)
Comment (idodod) [reply inputgraph/loader.go:469]: added a comment instead

File: inputgraph/loader.go [hunk 9]
```diff
@@ -454,6 +458,64 @@ func (l *loader) handleArg(ctx context.Context, cmd spec.Command, isBase bool) e
...
+}
+
+func (l *loader) handleSet(ctx context.Context, cmd spec.Command) error {
+	var opts commandflag.SetOpts
+	argsCpy := flagutil.GetArgsCopy(cmd)
+	args, err := flagutil.ParseArgsCleaned("LET", &opts, argsCpy)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse LET args")
+	}
+	if len(args) != 3 {
```
[the remark this answers is not in the document]
Comment (idodod) [reply inputgraph/loader.go:496]: see discussion above

**Comment by idodod** [comment]
> looking good; just a number of small suggestions.

Thanks for the review, I think I addressed all comments.

**Review by alexcb — APPROVED** [review]

**Pull request merged**